File: ainpc-core-plugin/src/main/java/ro/ainpc/progression/ProgressionObjectiveSnapshot.java
package ro.ainpc.progression;

import ro.ainpc.engine.ScenarioEngine;

public record ProgressionObjectiveSnapshot(
    String key,
    String type,
    String label,
    String description,
    String stageId,
    String stageLabel,
    String stateId,
    String stateDisplay,
    int currentAmount,
    int requiredAmount,
    boolean complete,
    boolean active
) {
    public ProgressionObjectiveSnapshot {
        key = valueOrEmpty(key);
        type = valueOrEmpty(type);
        label = valueOrEmpty(label);
        description = valueOrEmpty(description);
        stageId = valueOrEmpty(stageId);
        stageLabel = valueOrEmpty(stageLabel);
        stateId = valueOrEmpty(stateId);
        stateDisplay = valueOrEmpty(stateDisplay);
        currentAmount = Math.max(0, currentAmount);
        requiredAmount = Math.max(1, requiredAmount);
    }

    public static ProgressionObjectiveSnapshot fromQuestGuiObjective(ScenarioEngine.QuestGuiObjective objective) {
        if (objective == null) {
            return new ProgressionObjectiveSnapshot("", "", "", "", "", "", "", "", 0, 1, false, false);
        }

        return new ProgressionObjectiveSnapshot(
            objective.key(),
            objective.type(),
            objective.label(),
            objective.description(),
            objective.stageId(),
            objective.stageLabel(),
            objective.stateId(),
            objective.stateDisplay(),
            objective.currentAmount(),
            objective.requiredAmount(),
            objective.complete(),
            objective.active()
        );
    }

    private static String valueOrEmpty(String value) {
        return value == null ? "" : value.trim();
    }
}

File: ainpc-core-plugin/bin/main/ro/ainpc/progression/ProgressionObjectiveSnapshot.kt
package ro.ainpc.progression

import ro.ainpc.engine.ScenarioEngine

class ProgressionObjectiveSnapshot(
    key: String?,
    type: String?,
    label: String?,
    description: String?,
    stageId: String?,
    stageLabel: String?,
    stateId: String?,
    stateDisplay: String?,
    currentAmount: Int,
    requiredAmount: Int,
    private val completeValue: Boolean,
    private val activeValue: Boolean
) {
    private val keyValue: String = valueOrEmpty(key)
    private val typeValue: String = valueOrEmpty(type)
    private val labelValue: String = valueOrEmpty(label)
    private val descriptionValue: String = valueOrEmpty(description)
    private val stageIdValue: String = valueOrEmpty(stageId)
    private val stageLabelValue: String = valueOrEmpty(stageLabel)
    private val stateIdValue: String = valueOrEmpty(stateId)
    private val stateDisplayValue: String = valueOrEmpty(stateDisplay)
    private val currentAmountValue: Int = currentAmount.coerceAtLeast(0)
    private val requiredAmountValue: Int = requiredAmount.coerceAtLeast(1)

    fun key(): String = keyValue
    fun type(): String = typeValue
    fun label(): String = labelValue
    fun description(): String = descriptionValue
    fun stageId(): String = stageIdValue
    fun stageLabel(): String = stageLabelValue
    fun stateId(): String = stateIdValue
    fun stateDisplay(): String = stateDisplayValue
    fun currentAmount(): Int = currentAmountValue
    fun requiredAmount(): Int = requiredAmountValue
    fun complete(): Boolean = completeValue
    fun active(): Boolean = activeValue

    companion object {
        @JvmStatic
        fun fromQuestGuiObjective(objective: ScenarioEngine.QuestGuiObjective?): ProgressionObjectiveSnapshot {
            if (objective == null) {
                return ProgressionObjectiveSnapshot("", "", "", "", "", "", "", "", 0, 1, false, false)
            }
            return ProgressionObjectiveSnapshot(
                objective.key(),
                objective.type(),
                objective.label(),
                objective.description(),
                objective.stageId(),
                objective.stageLabel(),
                objective.stateId(),
                objective.stateDisplay(),
                objective.currentAmount(),
                objective.requiredAmount(),
                objective.complete(),
                objective.active()
            )
        }

        private fun valueOrEmpty(value: String?): String = value?.trim().orEmpty()
    }
}
